omega/omega-transaction/src/main/java/org/apache/servicecomb/pack/omega/transaction/RecoveryPolicyTimeoutWrapper.java

```diff
@@ -0,0 +1,214 @@
+      throws Throwable {
+    final TimeoutProb timeoutProb = new TimeoutProb(joinPoint, compensable);
+    this.timeoutProbs.add(timeoutProb);
+    Object output;
+    try {
+      output = this.recoveryPolicy
+          .applyTo(joinPoint, compensable, interceptor, context, parentTxId, retries);
+      if (timeoutProb.getInterruptFailureException() != null) {
+        throw new OmegaException(timeoutProb.interruptFailureException);
+      }
+    } catch (InterruptedException e) {
+      throw new TransactionTimeoutException(e.getMessage(),e);
+    } catch (IllegalMonitorStateException e) {
+      throw new TransactionTimeoutException(e.getMessage(),e);
+    } catch (ClosedByInterruptException e) {
+      throw new TransactionTimeoutException(e.getMessage(),e);
+    } catch (Throwable e) {
+      throw e;
+    } finally {
+      this.timeoutProbs.remove(timeoutProb);
+    }
+    return output;
+  }
+
+  /**
+   * Define timeout probe
+   */
+  private static final class TimeoutProb implements
+      Comparable<TimeoutProb> {
+
+    private final transient Thread thread = Thread.currentThread();
+    private final transient long startTime = System.currentTimeMillis();
+    private final transient long expireTime;
+    private Exception interruptFailureException = null;
+    private final transient ProceedingJoinPoint joinPoint;
+
+    public TimeoutProb(final ProceedingJoinPoint pnt, Compensable compensable) {
+      this.joinPoint = pnt;
+      this.expireTime = this.startTime + TimeUnit.SECONDS.toMillis(compensable.timeout());
+    }
+
+    @Override
+    public int compareTo(final TimeoutProb obj) {
+      int compare;
+      if (this.expireTime > obj.expireTime) {
+        compare = 1;
+      } else if (this.expireTime < obj.expireTime) {
+        compare = -1;
+      } else {
+        compare = 0;
+      }
+      return compare;
+    }
+
+    public Exception getInterruptFailureException() {
+      return interruptFailureException;
+    }
+
+    /**
+     *
+     * @return Returns TRUE if expired
+     */
+    public boolean expired() {
+      return this.expireTime < System.currentTimeMillis();
+    }
+
+    /**
+     * Interrupt thread
+     *
+     * @return Returns TRUE if the thread has been interrupted
+     */
+    public boolean interrupted() {
+      boolean interrupted;
+      if (this.thread.isAlive()) {
+        // 如果当前线程是活动状态，则发送线程中断信号
+        try {
+          this.thread.interrupt();
+        } catch (Exception e) {
+          this.interruptFailureException = e;
+          LOG.info("Failed to interrupt the thread " + this.thread.getName(), e);
+          throw e;
+        }
+        final Method method = MethodSignature.class.cast(this.joinPoint.getSignature()).getMethod();
+        LOG.warn("{}: interrupted on {}ms timeout (over {}ms)",
+            new Object[]{method, System.currentTimeMillis() - this.startTime,
+                this.expireTime - this.startTime}
+        );
+        interrupted = false;
```

Comment:
Is this value right?  I think it should true, as the thread is interrupted.

Reply:
This is OK because interrupt does not interrupt threads immediately, and `if (this. thread. isAlive ())` it needs to be checked again.